Refactor block_cemented_callback for readability

gr0vity-dev opened the pull request:
- Modularized the block confirmation logic into separate functions.
- Separated active and inactive block confirmation processes.
- Simplified the main block_cemented_callback() function for easier understanding.
- Reduced nested conditionals by early returning and using helper methods.
- Improved code readability and maintainability.

Marked as draft PR to discuss if such a breakup into smaller methods is desirable

gr0vity-dev committed:
Refactor block_cemented_callback for readability

- Modularized the block confirmation logic into separate functions.
- Separated active and inactive block confirmation processes.
- Simplified the main block_cemented_callback() function for easier understanding.
- Reduced nested conditionals by early returning and using helper methods.
- Improved code readability and maintainability.

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
 void nano::active_transactions::block_cemented_callback (std::shared_ptr<nano::block> const & block_a)
 {
-	auto transaction = node.store.tx_begin_read ();
+	nano::store::read_transaction transaction = node.store.tx_begin_read ();
```
Is using an explicit type intentional here? I think in most places we create transactions we use 'auto'

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
+	handle_final_votes_confirmation (block_a, transaction, *status_type);
+}
+
+boost::optional<nano::election_status_type> nano::active_transactions::determine_election_status (nano::store::read_transaction const & transaction, std::shared_ptr<nano::block> const & block)
```
I'd just call this "election_status"

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
+	if (!status_type)
+		return;
+
+	if (*status_type == nano::election_status_type::inactive_confirmation_height)
```
I think a switch statement would be better here, generally better for any enum value. Then the compiler can warn if there are missing cases which is important when we add new values to enums.

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
+void nano::active_transactions::process_inactive_confirmation (nano::store::read_transaction const & transaction, std::shared_ptr<nano::block> const & block)
+{
+	nano::account account{};
+	nano::uint128_t amount (0);
```
This should use initialization syntax:
nano::uint128_t amount{ 0 };

The way it's currently written is many other places in the code but it's an old style that's been replaced with braced initialization.

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
+
+void nano::active_transactions::process_inactive_confirmation (nano::store::read_transaction const & transaction, std::shared_ptr<nano::block> const & block)
+{
+	nano::account account{};
```
The {} is mostly redundant. For some types this forces the default constructor to be run, which could give it an initialized value versus an uninitialized one.

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
+	}
+}
+
+void nano::active_transactions::activate_scheduler_for_account_and_destination (const nano::account & account, std::shared_ptr<nano::block> const & block, nano::store::read_transaction const & transaction)
```
This name is a bit long. Maybe "activate_dependents".

gr0vity-dev replied:
I agree that it's too long. Since it activates the next blocks, what about "activate_successors" ?

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -67,99 +67,152 @@ void nano::active_transactions::stop ()
+	}
+}
+
+void nano::active_transactions::execute_confirmed (nano::store::read_transaction const & transaction, std::shared_ptr<nano::block> const & block, std::shared_ptr<nano::election> election, nano::election_status_type status_type)
```
Maybe "handle_confirmation" to maintain the "handle" nomenclature.

clemahieu reviewed (commented):
I added a couple style comments but since this is just moving code let's ignore those for now and do the direct code copying.

gr0vity-dev committed:
apply feedback

- auto transaction
- switch statement
- rename to `election_status`
- remove redundant {}
- use initialization syntax amount{ 0 }
- rename to `handle_confirmation`
- rename to `activate_successors`

gr0vity-dev commented:
Thanks for the feedback. I applied all your suggestions.
One thing I noticed is that we mostly use {variable}_a in the method definition, to indicate it's an argument ?
Should I keep that convention when splitting into smaller methods ?

clemahieu commented:
> Thanks for the feedback. I applied all your suggestions. One thing I noticed is that we mostly use {variable}_a in the method definition, to indicate it's an argument ? Should I keep that convention when splitting into smaller methods ?

I've been moving away from that convention. Since argument names are looked at by the users I think it's better to keep those clean.

I still use _m (class member) and _l (local variable) when there are name shadowing issues because those aren't often looked at by end users.

clemahieu commented on nano/node/active_transactions.cpp:
```diff
@@ -68,98 +68,155 @@ void nano::active_transactions::stop ()
+			process_inactive_confirmation (transaction, block_a);
+			break;
+
+		default:
```
I don't think we should use a default clause here but that's the existing functionality so I guess we can keep it.
